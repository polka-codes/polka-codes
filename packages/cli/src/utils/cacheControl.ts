import type { ModelMessage } from 'ai'

// Cache control configuration
const CACHEABLE_MODELS = ['sonnet', 'opus', 'haiku', 'gemini']

/**
 * Checks if a model supports caching based on model ID
 */
function isCacheableModel(modelId: string): boolean {
  return CACHEABLE_MODELS.some((model) => modelId.includes(model))
}

/**
 * Determines the provider key for cache control based on provider string
 */
function getProviderKey(provider: string): 'anthropic' | 'openrouter' | undefined {
  if (provider === 'openrouter') {
    return 'openrouter'
  }
  if (provider.includes('anthropic')) {
    return 'anthropic'
  }
  return undefined
}

/**
 * Applies cache control to messages for Anthropic and OpenRouter providers.
 * Adds ephemeral cache control to the last 2 user messages and system message.
 */
export function applyCacheControl(messages: ModelMessage[], provider: string, modelId: string): ModelMessage[] {
  const providerKey = getProviderKey(provider)

  if (!providerKey || !isCacheableModel(modelId)) {
    return messages
  }

  const providerOptions = { [providerKey]: { cacheControl: { type: 'ephemeral' } } }

  const newMessages = messages.slice()
  let userMessagesToUpdate = 2

  // Iterate backwards to find the system message and the last two user messages
  for (let i = newMessages.length - 1; i >= 0; i--) {
    const message = newMessages[i]
    const isUser = message.role === 'user'
    const isSystem = message.role === 'system'

    if ((isUser && userMessagesToUpdate > 0) || isSystem) {
      newMessages[i] = {
        ...message,
        providerOptions: {
          ...message.providerOptions,
          [providerKey]: {
            ...((message.providerOptions?.[providerKey] as Record<string, any>) ?? {}),
            ...providerOptions[providerKey],
          },
        },
      }

      if (isUser) {
        userMessagesToUpdate--
      } else if (isSystem) {
        break // Stop after updating the last system message
      }
    }
  }

  return newMessages
}
