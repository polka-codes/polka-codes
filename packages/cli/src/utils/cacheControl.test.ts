import { describe, expect, test } from 'bun:test'
import type { ModelMessage } from 'ai'
import { applyCacheControl } from './cacheControl'

describe('applyCacheControl', () => {
  const baseMessages: ModelMessage[] = [
    { role: 'system', content: 'System message' },
    { role: 'user', content: 'User message 1' },
    { role: 'assistant', content: 'Assistant message 1' },
    { role: 'user', content: 'User message 2' },
    { role: 'assistant', content: 'Assistant message 2' },
    { role: 'user', content: 'User message 3' },
  ]

  test('should not apply cache control for non-cacheable models', () => {
    const messages = applyCacheControl(baseMessages, 'anthropic', 'claude-2')
    expect(messages).toEqual(baseMessages)
    expect(messages).toMatchSnapshot()
  })

  test('should not apply cache control for unsupported providers', () => {
    const messages = applyCacheControl(baseMessages, 'openai', 'sonnet')
    expect(messages).toEqual(baseMessages)
    expect(messages).toMatchSnapshot()
  })

  test('should apply cache control for anthropic provider with a supported model', () => {
    const messages = applyCacheControl(baseMessages, 'anthropic', 'claude-3-sonnet-20240229')
    expect(messages).not.toEqual(baseMessages)
    expect(messages).toMatchSnapshot()
  })

  test('should apply cache control for openrouter provider with a supported model', () => {
    const messages = applyCacheControl(baseMessages, 'openrouter', 'anthropic/claude-3-opus')
    expect(messages).not.toEqual(baseMessages)
    expect(messages).toMatchSnapshot()
  })

  test('should apply cache control for gemini models', () => {
    const messages = applyCacheControl(baseMessages, 'openrouter', 'google/gemini-pro')
    expect(messages).not.toEqual(baseMessages)
    expect(messages).toMatchSnapshot()
  })

  test('should handle fewer than two user messages', () => {
    const shortMessages: ModelMessage[] = [
      { role: 'system', content: 'System message' },
      { role: 'user', content: 'User message 1' },
    ]
    const messages = applyCacheControl(shortMessages, 'anthropic', 'claude-3-sonnet-20240229')
    expect(messages).not.toEqual(shortMessages)
    expect(messages).toMatchSnapshot()
  })

  test('should work correctly without a system message', () => {
    const noSystemMessages: ModelMessage[] = [
      { role: 'user', content: 'User message 1' },
      { role: 'assistant', content: 'Assistant message 1' },
      { role: 'user', content: 'User message 2' },
      { role: 'user', content: 'User message 3' },
    ]
    const messages = applyCacheControl(noSystemMessages, 'anthropic', 'claude-3-sonnet-20240229')
    expect(messages).not.toEqual(noSystemMessages)
    expect(messages).toMatchSnapshot()
  })

  test('should merge with existing providerOptions', () => {
    const messagesWithProviderOptions: ModelMessage[] = [
      { role: 'system', content: 'System message' },
      { role: 'user', content: 'User message 1' },
      { role: 'assistant', content: 'Assistant message 1' },
      {
        role: 'user',
        content: 'User message 2',
        providerOptions: { anthropic: { someOtherOption: 'value' } },
      },
      { role: 'user', content: 'User message 3' },
    ]
    const messages = applyCacheControl(messagesWithProviderOptions, 'anthropic', 'claude-3-sonnet-20240229')
    expect(messages).not.toEqual(messagesWithProviderOptions)
    expect(messages[3].providerOptions?.anthropic).toHaveProperty('someOtherOption', 'value')
    expect(messages[3].providerOptions?.anthropic).toHaveProperty('cacheControl')
    expect(messages).toMatchSnapshot()
  })
})
