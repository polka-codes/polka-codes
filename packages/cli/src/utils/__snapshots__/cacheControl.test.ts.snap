// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`applyCacheControl should not apply cache control for non-cacheable models 1`] = `
[
  {
    "content": "System message",
    "role": "system",
  },
  {
    "content": "User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "role": "user",
  },
  {
    "content": "Assistant message 2",
    "role": "assistant",
  },
  {
    "content": "User message 3",
    "role": "user",
  },
]
`;

exports[`applyCacheControl should not apply cache control for unsupported providers 1`] = `
[
  {
    "content": "System message",
    "role": "system",
  },
  {
    "content": "User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "role": "user",
  },
  {
    "content": "Assistant message 2",
    "role": "assistant",
  },
  {
    "content": "User message 3",
    "role": "user",
  },
]
`;

exports[`applyCacheControl should apply cache control for anthropic provider with a supported model 1`] = `
[
  {
    "content": "System message",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "system",
  },
  {
    "content": "User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
  {
    "content": "Assistant message 2",
    "role": "assistant",
  },
  {
    "content": "User message 3",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
]
`;

exports[`applyCacheControl should apply cache control for openrouter provider with a supported model 1`] = `
[
  {
    "content": "System message",
    "providerOptions": {
      "openrouter": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "system",
  },
  {
    "content": "User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "providerOptions": {
      "openrouter": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
  {
    "content": "Assistant message 2",
    "role": "assistant",
  },
  {
    "content": "User message 3",
    "providerOptions": {
      "openrouter": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
]
`;

exports[`applyCacheControl should apply cache control for gemini models 1`] = `
[
  {
    "content": "System message",
    "providerOptions": {
      "openrouter": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "system",
  },
  {
    "content": "User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "providerOptions": {
      "openrouter": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
  {
    "content": "Assistant message 2",
    "role": "assistant",
  },
  {
    "content": "User message 3",
    "providerOptions": {
      "openrouter": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
]
`;

exports[`applyCacheControl should handle fewer than two user messages 1`] = `
[
  {
    "content": "System message",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "system",
  },
  {
    "content": "User message 1",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
]
`;

exports[`applyCacheControl should work correctly without a system message 1`] = `
[
  {
    "content": "User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
  {
    "content": "User message 3",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
]
`;

exports[`applyCacheControl should merge with existing providerOptions 1`] = `
[
  {
    "content": "System message",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "system",
  },
  {
    "content": "User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
        "someOtherOption": "value",
      },
    },
    "role": "user",
  },
  {
    "content": "User message 3",
    "providerOptions": {
      "anthropic": {
        "cacheControl": {
          "type": "ephemeral",
        },
      },
    },
    "role": "user",
  },
]
`;
